### PR TITLE
[libc++] Remove macros for keeping std::allocator members and void specialization after C++20

### LINKED ISSUE #75975: [libc++] Remove the _LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS escape hatch
As described in https://github.com/llvm/llvm-project/issues/69994, using the escape hatch makes us non-conforming in C++20. There are other problems like https://github.com/llvm/llvm-project/issues/63900.

We discussed this in the libc++ monthly meeting today and agreed that we should deprecate the macro in LLVM 18 and remove it in LLVM 19.

### PR DESCRIPTION
**Pull request by ilya-biryukov**
Fixes #75975.

Remove `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS` for the LLVM 19 release, it was previously marked as deprecated in LLVM 18.

I believe that `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_VOID_SPECIALIZATION` was only used by Google in conjunction with
`_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_MEMBERS`.

Removing both macros together should not cause any issues in practice, even though we did not announce the removal of
`_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_VOID_SPECIALIZATION` before.

### CODE REVIEW

**Comment by ilya-biryukov** [comment]
Please let me know if removal of `_LIBCPP_ENABLE_CXX20_REMOVED_ALLOCATOR_VOID_SPECIALIZATION` in this patch is controversial. I can see why someone would want to go with a more conservative approach (e.g. deprecate it in LLVM 19 and remove in LLVM 20), although my proposal would be to remove it right away. Happy to update this PR to go the other way.

I would also like to get feedback on the test updates as I wasn't sure I was taking all the right turns there.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx2a.verify.cpp [hunk 1]
```diff
@@ -12,12 +12,8 @@
 // pointer address(reference x) const;
```
Comment (philnik777) [inline libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx2a.verify.cpp:12]: Might as well rename the file to `address.cxx20.verify.cpp`.

**Review by philnik777 — COMMENTED** [review]
LGTM, but let's wait for Louis.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/allocate.cxx2a.pass.cpp [hunk 2]
```diff
@@ -1,93 +0,0 @@
-//===----------------------------------------------------------------------===//
```
Comment (ldionne) [inline libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/allocate.cxx2a.pass.cpp:1]: I am a bit confused -- why are we removing this test? The name says `allocate.cxx2a.pass.cpp` but it seems that this test was running in all configurations and was our only test for `allocate(size_t, void const*)` in e.g. C++11. Did I miss something?

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator_types.cxx2a.pass.cpp [hunk 3]
```diff
@@ -1,51 +0,0 @@
-//===----------------------------------------------------------------------===//
```
Comment (ldionne) [inline libcxx/test/libcxx/depr/depr.default.allocator/allocator_types.cxx2a.pass.cpp:1]: I have the same question here, I don't understand why this is removed in all standard modes.

**Review by ldionne — COMMENTED** [review]
Thanks a lot for picking up this task, I really appreciate it! I have some questions but it's probably just me missing the point, this patch looks good overall.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/allocate.cxx2a.pass.cpp (hunk 2 above)
Comment (ilya-biryukov) [reply libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/allocate.cxx2a.pass.cpp]: I initially thought that it was checking the allocator members keep working in C++20 when the corresponding macro is defined (and running in other standard modes for sanity checking or simply redundantly as an oversight). I implicitly assumed there are other tests checking these members in pre-C++20 modes.

But I was wrong, thanks for pointing this out, it was the only test that has the deprecated version of allocate with hint, I've restored it and made it run only for pre-C++20 mode.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx2a.verify.cpp (hunk 1 above)
Comment (ilya-biryukov) [reply libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx2a.verify.cpp]: Done. Renamed other cxx2a files in the directory as well.

PS I'd normally land things like this as a separate `NFC` commit to keep the actual changes small, but I'm just doing it all in a single commit here. Let me know if you'd also prefer that to be a separate follow-up change.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator_types.cxx2a.pass.cpp (hunk 3 above)
Comment (ilya-biryukov) [reply libcxx/test/libcxx/depr/depr.default.allocator/allocator_types.cxx2a.pass.cpp]: Restored the file, changed it to run in pre-C++20 modes (and all the other deleted files too)

**Comment by ilya-biryukov** [comment]
> Thanks a lot for picking up this task, I really appreciate it! I have some questions but it's probably just me missing the point, this patch looks good overall.

Thanks for a quick review and an insightful comment. I think I did remove some useful tests, it was unintentional, I had some wrong assumptions about the tests that I didn't check. They should be restored now.

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx20.verify.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
+//===----------------------------------------------------------------------===//
```
Comment (ldionne) [inline libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx20.verify.cpp:1]: Can you clang-format your patch?

**Review by ldionne — APPROVED** [review]
LGTM w/ green CI and clang-format. Thanks!

File: libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx20.verify.cpp (hunk 4 above)
Comment (ilya-biryukov) [reply libcxx/test/libcxx/depr/depr.default.allocator/allocator.members/address.cxx20.verify.cpp:1]: Done.

**Pull request merged**